src/vm/i386/stublinkerx86.cpp

```diff
@@ -5930,11 +5930,28 @@ static void AppendGCLayout(ULONGARRAY &gcLayout, size_t baseOffset, BOOL fIsType
         {
             FindByRefPointerOffsetsInByRefLikeObject(
                 pMT,
-                baseOffset,
+                0 /* baseOffset */,
                 [&](size_t pointerOffset)
                 {
-                    _ASSERTE(gcLayout.Count() == 0 || pointerOffset > (gcLayout[gcLayout.Count() - 1] & ~(ULONG)1));
-                    *gcLayout.AppendThrowing() = (ULONG)(pointerOffset | 1); // "| 1" to mark it as an interior pointer
+                    // 'gcLayout' requires stack offsets relative to the top of the stack to be recorded, such that subtracting
+                    // the offset from the stack top yields the address of the field, given that subtracting 'baseOffset' from
+                    // the stack top yields the address of the first field in this struct. See TailCallFrame::GcScanRoots() for
+                    // how these offsets are used to calculate stack addresses for fields.
+                    _ASSERTE(pointerOffset < baseOffset);
+                    size_t stackOffsetFromTop = baseOffset - pointerOffset;
+                    _ASSERTE(FitsInU4(stackOffsetFromTop));
+
+                    // Offsets in 'gcLayout' are expected to be in increasing order
+                    int gcLayoutInsertIndex = gcLayout.Count();
+                    _ASSERTE(gcLayoutInsertIndex >= 0);
+                    while (gcLayoutInsertIndex != 0 && stackOffsetFromTop < gcLayout[gcLayoutInsertIndex - 1])
+                    {
+                        --gcLayoutInsertIndex;
+                        _ASSERTE(stackOffsetFromTop != (gcLayout[gcLayoutInsertIndex] & ~(ULONG)1));
+                    }
+
+                    _ASSERTE(gcLayout.Count() == 0 || stackOffsetFromTop > (gcLayout[gcLayout.Count() - 1] & ~(ULONG)1));
```

Comment:
I think this assert will fire if somebody creates struct with overlapping Spans:
```
[StructLayout(LayoutKind.Explicit)]
ref struct MyStruct
{
    [FieldOffset(0)]
    Span<int> _span1;
    [FieldOffset(0)]
    Span<int> _span2;
}
```

Reply:
Yep it fires, added a check to see if the offset was already added